### ISSUE TITLE
bosh-dns-alias recursing issue

### ISSUE THREAD

**Issue by poblin-orange** [orange-cloudfoundry/paas-templates#1011]
### Expected behavior

* As a paas-templates operator 
* In order to 
  - benefit from local (bosh deployment level) dns resolution
* I need 
  - bosh-dns-aliases set locally to be used before bosh-dns agent tries to recurse to paas-templates global dns (192.168.116.xx)

### Observed behavior
- bosh-dns-alias config file is populated
```
server/18685722-6cc3-4b6c-812c-804b9a5ab3a2:~# cat /var/vcap/jobs/bosh-dns-aliases/dns/aliases.json 
{
...
  "postgres": [
    "*.postgres.net-bosh-2.00-gitops-management.bosh"
  ]
}
``` 

- bosh-dns agent configures /etc/resolv.conf
```
# cat /etc/resolv.conf 
# This file was automatically updated by bosh-dns
nameserver 169.254.0.2

nameserver 192.168.116.156
nameserver 192.168.116.166
```
- bosh-dns start (postgres alias is registred)
```
server/18685722-6cc3-4b6c-812c-804b9a5ab3a2:/var/vcap/sys/log/bosh-dns# tail -f bosh_dns.stdout.log
[FailoverRecursor] 2020-11-17T17:12:45.296101000Z INFO - starting preference: 192.168.116.156:53
[HandlerRegistrar] 2020-11-17T17:12:45.296223000Z INFO - Register bosh. as internal domain
[HandlerRegistrar] 2020-11-17T17:12:45.296230000Z INFO - Register cf-deployment-routers. as internal domain
[HandlerRegistrar] 2020-11-17T17:12:45.296236000Z INFO - Register cf-deployment-schedulers. as internal domain
[HandlerRegistrar] 2020-11-17T17:12:45.296242000Z INFO - Register concourse-db.internal.paas. as internal domain
[HandlerRegistrar] 2020-11-17T17:12:45.296249000Z INFO - Register jumpbox. as internal domain
[HandlerRegistrar] 2020-11-17T17:12:45.296256000Z INFO - Register ops-router. as internal domain
[HandlerRegistrar] 2020-11-17T17:12:45.296263000Z INFO - Register ops-routing-nats. as internal domain
[HandlerRegistrar] 2020-11-17T17:12:45.296281000Z INFO - Register postgres. as internal domain
```

```
# dig postgres @169.254.0.2

; <<>> DiG 9.10.3-P4-Ubuntu <<>> postgres @169.254.0.2
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 50338
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;postgres.			IN	A

;; ANSWER SECTION:
postgres.		0	IN	A	192.168.116.218

;; Query time: 0 msec
;; SERVER: 169.254.0.2#53(169.254.0.2)
;; WHEN: Tue Nov 17 17:50:13 UTC 2020
;; MSG SIZE  rcvd: 61
```

- tcpdumping effective request
```
server/18685722-6cc3-4b6c-812c-804b9a5ab3a2:~# tcpdump -nttti eth0 dst port 53
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
 00:00:00.000000 IP 192.168.116.219.44915 > 192.168.116.156.53: 57753+ A? postgres. (26)
 00:00:00.000014 IP 192.168.116.219.44915 > 192.168.116.166.53: 57753+ A? postgres. (26)
 00:00:00.000008 IP 192.168.116.219.44915 > 192.168.116.156.53: 57976+ AAAA? postgres. (26)
 00:00:00.000005 IP 192.168.116.219.44915 > 192.168.116.166.53: 57976+ AAAA? postgres. (26)

```
==> seems the bosh-dns does not always resolve locally, resulting in dns lookup failure under load.

==> hardcoding the postgres alias manually in /etc/hosts fixes the lookup issue (not long, as bosh-dns rewrites it each time the director issues a dns_sync order)

==> misconfiguration or bug in bosh-dns job

### references
- https://bosh.io/docs/dns/#monitoring
- 
 

### Affected releases

*  48.0.2
* and earlier versions

**Comment by gberche-orange** [comment]
#905 might help monitor DNS resolution times in bosh dns

**Comment by poblin-orange** [comment]
bump release 1.27.0 =>  same issue

**Comment by poblin-orange** [comment]
Workaround: use direct bosh-dns syntax in the job propertie ie:
```
0.mysql.net-bosh-2.00-gitops-management.bosh
`` 
=> no more dns resolution issue

**Comment by poblin-orange** [comment]
Same issue on:
- k3s release 17 1.20.0
- bosh-dns 1.29.0
- paas-templates 51.0.3
- traefik externalName mapping to:
  - local bosh-dns-alias release alias configuration
  - director level addon on bosh-dns release  alias confgururation

Workaround by using direct q-s0.xxxx alias in k8s ExternalName services

Collected logs

**Comment by gberche-orange** [comment]
Problem was observed during v52 upgrade pipeline on the vsphere integration plateform.

Root cause seems the core-dns forward plugin which is using all entries from specified `resolv.conf` backend in random order.

See https://github.com/coredns/coredns/issues/3900

This regression was introduced in core-dns 1.6.7 and [fixed in core-dns 1.7.0](https://coredns.io/2020/06/15/coredns-1.7.0-release/)
